[core][runtime_env] Allow full path in conda runtime env.

rynewang opened the pull request:
#34956 fixed a conda hanging issue by checking existence of the conda env name. However it only checks by conda env name, not path - if the user specifies an absolute path, it no longer works. This is an regression because it should support full paths without a problem.

This PR reads conda CLI for the full paths and names. A special treatment is done for `base` which is just the conda base prefix; and others are base conda_prefix/envs/name.

Adds tests for both `base` and full path of it; and also name and full paths for non base envs.

Fixes #44373 .

jjyao commented:
```
conda (dict | str): Either (1) a dict representing the conda environment YAML, (2) a string containing the path to a local [conda “environment.yml”](https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#create-env-file-manually) file, or (3) the name of a local conda environment already installed on each node in your cluster (e.g., "pytorch_p36")
```

from https://docs.ray.io/en/latest/ray-core/handling-dependencies.html doesn't mention that we support full path only the name.


If we want to support it, we should also update our doc.

rynewang commented:
Doc updated

rynewang commented:
@angelinalg : added a clause to the doc about how to specify `conda` runtime env. Previously you can set a env name like `pytorch_p36`, now you can also set an absolute path of it like `"/home/youruser/anaconda3/envs/pytorch_p36"`. Also updated some single backticks to double backticks to be consistent with rest of doc.

hongchaodeng commented on python/ray/_private/runtime_env/conda_utils.py:
```diff
@@ -173,6 +173,33 @@ def get_conda_env_list() -> list:
     return envs
 
 
+def get_conda_info_json() -> dict:
```
Can you add a link or an example of the return value?

rynewang replied:
done

hongchaodeng reviewed (approved):
LGTM after nit
